Mathlib/SetTheory/Ordinal/Nimber.lean

```diff
@@ -0,0 +1,362 @@
+  rfl
+
+theorem succ_def (a : Nimber) : succ a = toNimber (toOrdinal a + 1) :=
+  rfl
+
+/-- A recursor for `Nimber`. Use as `induction x`. -/
+@[elab_as_elim, cases_eliminator, induction_eliminator]
+protected def rec {β : Nimber → Sort*} (h : ∀ a, β (toNimber a)) : ∀ a, β a := fun a =>
+  h (toOrdinal a)
+
+/-- `Ordinal.induction` but for `Nimber`. -/
+theorem induction {p : Nimber → Prop} : ∀ (i) (_ : ∀ j, (∀ k, k < j → p k) → p j), p i :=
+  Ordinal.induction
+
+protected theorem le_zero {a : Nimber} : a ≤ 0 ↔ a = 0 :=
+  Ordinal.le_zero
+
+protected theorem not_lt_zero (a : Nimber) : ¬ a < 0 :=
+  Ordinal.not_lt_zero a
+
+protected theorem pos_iff_ne_zero {a : Nimber} : 0 < a ↔ a ≠ 0 :=
+  Ordinal.pos_iff_ne_zero
+
+instance (a : Nimber.{u}) : Small.{u} (Set.Iio a) :=
+  Ordinal.small_Iio a
+
+instance (a : Nimber.{u}) : Small.{u} (Set.Iic a) :=
+  Ordinal.small_Iic a
+
+end Nimber
+
+theorem not_small_nimber : ¬ Small.{u} Nimber.{max u v} :=
+  not_small_ordinal
+
+namespace Ordinal
+
+variable {a b c : Ordinal.{u}}
+
+@[simp]
+theorem toNimber_symm_eq : toNimber.symm = Nimber.toOrdinal :=
+  rfl
+
+@[simp]
+theorem toNimber_toOrdinal (a : Ordinal) :  Nimber.toOrdinal (toNimber a) = a :=
+  rfl
+
+@[simp]
+theorem toNimber_zero : toNimber 0 = 0 :=
+  rfl
+
+@[simp]
+theorem toNimber_one : toNimber 1 = 1 :=
+  rfl
+
+@[simp]
+theorem toNimber_eq_zero (a) : toNimber a = 0 ↔ a = 0 :=
+  Iff.rfl
+
+@[simp]
+theorem toNimber_eq_one (a) : toNimber a = 1 ↔ a = 1 :=
+  Iff.rfl
+
+@[simp]
+theorem toNimber_max (a b : Ordinal) :
+    toNimber (max a b) = max (toNimber a) (toNimber b) :=
+  rfl
+
+@[simp]
+theorem toNimber_min (a b : Ordinal) :
+    toNimber (min a b) = min (toNimber a) (toNimber b) :=
+  rfl
+
+end Ordinal
+
+/-! ### Nimber addition -/
+
+namespace Nimber
+
+variable {a b c : Nimber.{u}}
+
+/-- Nimber addition is recursively defined so that `a + b` is the smallest number not equal to
+`a' + b` or `a + b'` for `a' < a` and `b' < b`. -/
+-- We write the binders like this so that the termination checker works.
+protected def add (a b : Nimber.{u}) : Nimber.{u} :=
+  sInf {x | (∃ a', ∃ (_ : a' < a), Nimber.add a' b = x) ∨
+    ∃ b', ∃ (_ : b' < b), Nimber.add a b' = x}ᶜ
+termination_by (a, b)
+
+instance : Add Nimber :=
+  ⟨Nimber.add⟩
+
+theorem add_def (a b : Nimber) :
+    a + b = sInf {x | (∃ a' < a, a' + b = x) ∨ ∃ b' < b, a + b' = x}ᶜ := by
+  change Nimber.add a b = _
+  rw [Nimber.add]
+  simp_rw [exists_prop]
+  rfl
+
+/-- The set in the definition of `Nimber.add` is nonempty. -/
+private theorem add_nonempty (a b : Nimber.{u}) :
+    {x | (∃ a' < a, a' + b = x) ∨ ∃ b' < b, a + b' = x}ᶜ.Nonempty := by
+  simp_rw [Set.nonempty_compl, Set.setOf_or, ← Set.mem_Iio, ← Set.image.eq_1]
+  apply_fun (fun a : Set Nimber ↦ Small.{u} a)
+  have : Small.{u} ↑((· + b) '' Set.Iio a ∪ (a + ·) '' Set.Iio b) := inferInstance
+  simp [this, small_congr (Equiv.Set.univ _)]
+  exact not_small_nimber
```

Comment:
```suggestion
  simpa [this, small_congr (Equiv.Set.univ _)] using not_small_nimber
```
looks like it might work, and eliminates a non-terminal simp.
If it doesn't work, then try to eliminate the non-terminal simp in some other way